drivers: mpsl: Add flash sync driver

eugmes opened the pull request:
This PR adds flash sync support for the `soc_flash_nrf` driver using MPSL timeslot API. It also removes the flash driver provided by `bt_ll_softdevice` and replaces it by `soc_flash_nrf` with MPSL sync enabled.

I'm creating the PR here so the review could start before upmerge.

tejlmand committed:
treewide: Update OSS projects revisions

sdk-zephyr: a4ead98051404e3b63845d9944b0c689162f1db1 (PR #354)
sdk-mcuboot: ec441e0a700280673e11f4e242ff5de696d4f962 (PR #120)
sdk-mcumgr: b3716c64bde95ea5911a36001fd8f8d4b46eb310 (PR #10)
sdk-nrfxlib: f1dd2ee510f39e27ea4102bad369c026c0dad147 (PR #266)

Summary of changes needed for this upmerge:

- modules: device: Const-ify all device driver instance pointers.

  See also https://github.com/zephyrproject-rtos/zephyr/pull/24873

- jenkins: remove --board-root zephyr/boards

  zephyr/boards are always part of sanity-check and with duplicate board
  checking with: https://github.com/zephyrproject-rtos/zephyr/pull/27549
  this now results in CI failures.

- samples: whitelist -> allow

  sanitycheck: inclusive language.

  change whitelist -> allow for all sample.yaml and testcase.yaml files.

- scripts: ncs_commands: add hal_quicklogic to _BLOCKED_PROJECTS

  This is a vendor HAL for an SoC family not supported in NCS.

- doc: upmerge changes to docs

  Adjusted `CONFIG_BT_GATT_DIS` -> `CONFIG_BT_DIS` in docs.

- samples: remove GATT infix in conf files for applications and samples

- usb: hid: All hid_ops callbacks get device pointer.

  Updated usb_state.c to new hid_ops callbacks which includes the device
  pointer.

- cmake: bintools elfconvert

  Updated according to new toolchain abstraction.

- partition_manager: cmake: update regex for runner hex file

  The format of the runner yaml has changed. Update the regex used to
  replace the hex_file variable to match it.

- spm: align with updated API for setting irq state

  The API for setting IRQ state has changed.
  Update to align with the new functionailty.

  Ref: NCSDK-6463

- nrf_desktop: using CONFIG_SYSTEM_CLOCK_WAIT_FOR_STABILITY=y

  Updated nrf_desktop applications to use upstream
  SYSTEM_CLOCK_WAIT_FOR_STABILITY instead of
  CLOCK_CONTROL_NRF_K32SRC_BLOCKING.

- dts/Kconfig: Remove HAS_DTS_I2C

  HAS_DTS_I2C is now selected by I2C and
  always used as I2C && HAS_DTS_I2C.

  It could then be purely removed.

- drivers: mpsl: clock_control: Provide nrfx_clock shim

  Replace clock_control_mpsl with nrfx_clock implementation which is
  using mpsl_clock API. This way standard clock_control driver is
  used and only thin, straightforward nrfx_clock API is replaced with
  implementation which is utilizing mpsl_clock API.

- mpsl: removing clock control in mpsl

- applications: nrf_desktop: Disable mem protection in mcuboot

  Disable HW stack protection in MCUBoot. This seems to cause
  problems as region protection migrates to application and can
  cause faults during kernel initialization.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>
Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>
Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>
Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

rugeGerritsen commented on subsys/mpsl/Kconfig:
```diff
@@ -21,6 +21,7 @@ config MPSL_SIGNAL_STACK_SIZE
 
 config MPSL_TIMESLOT_SESSION_COUNT
 	int "Number of timeslot sessions"
+	default 1 if SOC_FLASH_NRF_RADIO_SYNC_MPSL
```
Can we rather update `mpsl_init.c` to add 1 if flash driver is included?

eugmes replied:
Should the comment be changed to something like "Number of additional timeslot sessions"

rugeGerritsen commented on drivers/mpsl/flash_sync/flash_sync_mpsl.c:
```diff
@@ -0,0 +1,151 @@
+
+	__ASSERT_NO_MSG(k_sem_count_get(&_context.sem) == 0);
+
+	ret = mpsl_timeslot_request(_context.session_id, req);
```
Timeslot APIs are not threadsafe. Either we need to ensure that they are called from a cooperative thread, or we need to use the multithreading lock.

eugmes replied:
Which thread are they supposed to be called from?

eugmes replied:
Note also that flash driver calls are protected by a semaphore.

rugeGerritsen commented on drivers/mpsl/flash_sync/flash_sync_mpsl.c:
```diff
@@ -0,0 +1,151 @@
+	req->params.earliest.hfclk = MPSL_TIMESLOT_HFCLK_CFG_NO_GUARANTEE;
+	req->params.earliest.priority = MPSL_TIMESLOT_PRIORITY_NORMAL;
+	req->params.earliest.length_us = _context.duration;
+	req->params.earliest.timeout_us = MPSL_TIMESLOT_EARLIEST_TIMEOUT_MAX_US;
```
You may consider setting the timeout in a more intelligent way based on the timeout set of the flash context

eugmes replied:
I don't think that this warrants added complexity. That timeout should only be triggered if there is some serious error in the driver implementation.

rugeGerritsen replied:
If you specify a shorter timeout you will be able to get a failure immediately. Otherwise you may end up in a situation where the flash operation is scheduled far into the future, after the timeout set by the upper layer.

rugeGerritsen commented on drivers/mpsl/flash_sync/flash_sync_mpsl.c:
```diff
@@ -0,0 +1,151 @@
+}
+
+void nrf_flash_sync_get_timestamp_begin(void)
+{
```
You could add a comment "Not needed for the MPSL synchronization backend"

rugeGerritsen commented on drivers/mpsl/flash_sync/flash_sync_mpsl.c:
```diff
@@ -0,0 +1,151 @@
+	req->request_type = MPSL_TIMESLOT_REQ_TYPE_EARLIEST;
+	req->params.earliest.hfclk = MPSL_TIMESLOT_HFCLK_CFG_NO_GUARANTEE;
+	req->params.earliest.priority = MPSL_TIMESLOT_PRIORITY_NORMAL;
+	req->params.earliest.length_us = _context.duration;
```
From what I can see, `_context.duration` cannot be trusted for write operations. We need to do the calculations ourselves.

For writes, the upper layer sets the value 7500 which may or may not be sufficient.

Similar for erase, the timeout is set to the time required to erase all pages. This gives bad scheduling performance. 
For partial erasing, the time required is also multiplied by 2 as it takes context switching into account.

eugmes replied:
There is no way to do any calculations on this level. There is absolutely no information about the kind of operation that is preformed.

rugeGerritsen replied:
I see. I'm not sure how to come around this limitation. @nvlsianpu  do you have some input on how this can be solved without requiring to wait for the next upmerge?

eugmes replied:
I think `duration` is given for one page. The timeout for the complete set of operations is `FLASH_TIMEOUT_MS`.

rugeGerritsen commented on drivers/mpsl/flash_sync/flash_sync_mpsl.c:
```diff
@@ -0,0 +1,164 @@
+
+	case MPSL_TIMESLOT_SIGNAL_SESSION_IDLE:
+	case MPSL_TIMESLOT_SIGNAL_SESSION_CLOSED:
+	case MPSL_TIMESLOT_SIGNAL_CANCELLED:
```
MPSL_TIMESLOT_SIGNAL_CANCELLED should be handled the same way as MPSL_TIMESLOT_SIGNAL_BLOCKED

eugmes committed:
drivers: mpsl: Add flash sync driver

Add flash sync support for the soc_flash_nrf driver using MPSL
timeslot API.

Signed-off-by: Ievgenii Meshcheriakov <ievgenii.meshcheriakov@nordicsemi.no>

eugmes committed:
drivers: bt_ll_softdevice: Remove the flash driver

MPSL flash sync driver should be used instead together
with the normal NRF flash driver.

Signed-off-by: Ievgenii Meshcheriakov <ievgenii.meshcheriakov@nordicsemi.no>

rugeGerritsen commented on drivers/mpsl/flash_sync/Kconfig:
```diff
@@ -0,0 +1,30 @@
+
+endif
+
+config SOC_FLASH_NRF_RADIO_SYNC_MPSL_TIMESLOT_SESSION_COUNT
```
Do we need a separate config for this? The flash driver will always use one sesssion

eugmes replied:
Yes, but the code in `mpsl_init.c` is easier if this specifies 0 or 1.

rugeGerritsen replied:
I see your point. Still I would prefer to not have this config. It indicates to the reader that the value can be greater than 1

eugmes replied:
The reader would not be able to modify this anyway.

eugmes replied:
I mean modify using menuconfig or prj.conf.

Commit:
doc: fix key name while accessing sample.yaml

Fix table-from-rows extension to look for 'platform_allow', instead of
old 'platform_whitelist'.

Signed-off-by: Fabio Utzig <fabio.utzig@nordicsemi.no>

eugmes commented:
Review to continue at https://github.com/nrfconnect/sdk-nrf/pull/3003
